### ISSUE TITLE
Thoughts on Data Models, Interoperability, and Querying Trust Registries

### ISSUE THREAD

**Issue by mathieuglaude** [trustoverip/tswg-trust-registry-protocol#21]
When exploring the concept of trust registries, it becomes evident that the **data models** for each registry vary significantly. For instance, the data model used by the DIACC PCTF is markedly different from those of the C2PA, and similarly, the models for issuer and verifier registries, as well as membership registries, all present their unique structures.

This diversity in data models reflects the nature of trust as an inherently multifaceted concept. Each trust registry serves a distinct purpose and, therefore, can necessitate a specific type of data model. This leads to the realization that, while the registries are diverse, there is a need to categorize them into types. Just as we categorize professionals like plumbers, engineers, and doctors in a directory, trust registries could benefit from a similar classification system.

However, a challenge arises when we consider the global scope of these registries. A plumbers list in Ontario might have a different data format compared to one in Japan, which is acceptable. The critical question is whether we can acknowledge the existence of such a list, regardless of its format.

The concept of 'type' becomes crucial here. Each type should have an associated data model and a defined source of truth. This framework then raises the question of how to formulate queries across these diverse systems in a standardized manner.

For example, querying whether a person is authorized to issue a credential requires a different approach than inquiring if a company is a member of a mining association or if they are permitted to mine coal. These distinct queries highlight the need for a method to construct and ask questions effectively.

One potential solution involves leveraging artificial intelligence and generative techniques, along with expertise in regular expressions, to craft these queries. This approach is akin to using predicate proof requests, where the ability to ask a specific question, such as verifying if someone's age is over 18, is predicated on knowing the correct credential. Drawing a parallel, knowing the 'type' of registry and the data model it employs allows for precise and meaningful queries.

Note: I realize that this topic may be better suited in 'Discussions', however I think it impacts how the protocol should be designed. I can move it to there @darrellodonnell if you think its better suited for there..

**Comment by darrellodonnell** [comment]
I agree that data models vary wildly. I submit that the (simplistic) use of "namespaces" allows us to begin normalizing  the conversation.

If you have multiple countries looking at a range of different schema (e.g. driver license) you can create a namespace to help normalize - but you'll need to consider equivalencies and purpose to do that. As an example - the International Driver License normalizes a driver license to fit a "standard" set of attributes that answer the question "Is Bob licensed to drive in the US?" so decisions can be made elsewhere. 

But we also have "treaties" that link states to states and states to provinces for shared purposed (e.g. get a fine in some states and you will see demerit points on your Ontario driver license" - and those are on a case by base basis, established jurisdiction-to-jurisdiction...

We could easily have multiple namespaces supported by any data model:

* Does Bob have a `idl:equivalentdriverlicense` under Global DL Framework?
* Does Bob have a `ontario:driverlicense` under NY-ON-demerit-framework? Helps a NY police system connect the driver to a treaty-enabled system.

The work of namespacing (agreeing on a single term) is, IMO, easier than attempting to normalize full systems when you're looking at equivalency mapping for a discrete purpose. 


> For example, querying whether a person is authorized to issue a credential requires a different approach than inquiring if a company is a member of a mining association or if they are permitted to mine coal. These distinct queries highlight the need for a method to construct and ask questions effectively.

Do you see a single TR that answers these two very different questions:
* is EntityX a member of a mining association? 
* is EntityX permitted to mine coal?

Those are very different questions and to ask them without domain knowledge seems odd to me.

**Comment by mathieuglaude** [comment]
I think namespaces could be a good idea. But depending on the type of trust registry, it can answer multiple (bounded) questions specific to that type. A TR with authorized issuers may answer one or two questions. But a membership list type may answer 4-5.

So depending on the type, they may answer to a set of specific questions. So I guess certain questions must be agreed upon based on the type. But there may be multiple questions asked to one TR.

**Comment by darrellodonnell** [comment]
Exactly, and each ecosystem will look different but they will "rhyme".

**Comment by andorsk** [comment]
I’d like to revisit and revive this conversation.

After attending #IIW39, I’ve spent considerable time reflecting on the challenges of namespacing, particularly around the issue of "collisions." When namespaces are human-readable, it will likely result in collisions, which would create vocabulary confusion—especially when an actor operates across multiple ecosystems.

For example, in the question above: "is EntityX permitted to mine coal?", "mining" and "permitted" may mean different things depending on jurisdiction. 

To address this, I propose that all namespaces MUST be globally unique. If all namespaces are globally unique (i.e a DID, then it would prevent conflicts. Lookups on each unique namespace identifier could then provide additional metadata, offering more human-readable information when needed. We could refine the metadata model in time, which could, in theory, enable people understand vocabulary better. 

The tradeoff here is the following:

1. Either we lose human readability on the referencing identifier by the need for global uniqueness OR
2. We have to have a globally conflict free namespacing service  (i.e think about dns).

Here's a small illustration of where a conflict of namespaces may happen. 

![image](https://github.com/user-attachments/assets/c72629e4-e2e0-4db5-b875-1c19c3619c6a)

**Comment by andorsk** [comment]
We are working on an abstract data model for the TRQP in the meetings lately. This issue seems to align nicely with this work.

**Comment by andorsk** [comment]
Coming back to this work, we have a concrete answer here. In particular https://trustoverip.github.io/tswg-trust-registry-protocol/#problem-statement and https://trustoverip.github.io/tswg-trust-registry-protocol/#interpretation-of-the-diagram-ecosystem-and-trust-registry-relationship, https://trustoverip.github.io/tswg-trust-registry-protocol/#metadata-models  describe the relationships and data models 

![Image](https://github.com/user-attachments/assets/7c0e6a31-e7c2-404a-91a7-cca32db8faa4)

Suggest we. revisit this issue once all remaining PR's are merged, and make sure to see if we can close this off.

**Comment by andorsk** [comment]
I believe the spec's abstraction of the query methods solves this issue :  

> For example, querying whether a person is authorized to issue a credential requires a different approach than inquiring if a company is a member of a mining association or if they are permitted to mine coal. These distinct queries highlight the need for a method to construct and ask questions effectively.

### 1. **Is a person authorized to issue a credential?**
- **authority_id** → The Trust Registry responsible for authorizations (e.g., `tr_authorization`)
- **assertion** → `can_issue_credential` (or a standardized assertion like `authorization`)
- **entity_id** → The person's DID

**Resulting Query:**  
`description: tr_authorization, can_issue_credential, did:example:person123`

### 2. **Is a company a member of a mining association?**
- **authority_id** → The Mining Association’s Trust Registry (e.g., `tr_mining_association`)
- **assertion** → `membership`
- **entity_id** → The company’s DID

**Resulting Query:**  
`description: tr_mining_association, membership, did:example:company456`

### 3. **Is a company permitted to mine coal?**
- **authority_id** → The national or regional mining regulator (e.g., `tr_mining_regulator`)
- **assertion** → `can_mine_coal`
- **entity_id** → The company’s DID

**Resulting Query:**  
`description: tr_mining_regulator, can_mine_coal, did:example:company456`

**Comment by andorsk** [comment]
@mathieuglaude : motion to close if you feel this is appropriately addressed.

**Comment by andorsk** [comment]
Authority is not the EGF. The authority is the thing making the statement. In the case of a TR Provider, it's the TR provider. In the case of an ecosystme, it's the ecosystem authority. EGF's are linked, but a different concept.  

The model explain this. The authority_id MUST be a part of the query. It is possible to create an assumption and implicitly derive the authority ID based on context, but this is an implementation detail and there are possible risks with this approach that a network will have to decide.

**Comment by darrellodonnell** [comment]
@mathieuglaude could you spin this up as a Discussion item? I would like to reach closure on the Issues list as we move towards Public Review 02 this month. This topic is critically important to not lose though. Systems of Record and other sources are, to your point, wildly varied. TRQP serves to provide a simple way to get consistent answers on only a few questions (authorization or recognition primarily, IMO). The operators of these systems of record need to understand that TRQP doesn't as for much - if any - change to their systems.

**Comment by darrellodonnell** [comment]
moved to discussion with link back to this issue. 

https://github.com/trustoverip/tswg-trust-registry-protocol/discussions/147

CLOSING
